File: CoresRetro/RetroArch/ZipArchive/Example/ObjectiveCExample/ViewController.m
//
//  ViewController.m
//  ObjectiveCExample
//
//  Created by Sean Soper on 10/23/15.
//
//

#import "ViewController.h"

#if COCOAPODS
#import <SSZipArchive.h>
#else
#import <ZipArchive.h>
#endif


@interface ViewController ()

@property (weak, nonatomic) IBOutlet UITextField *passwordField;
@property (weak, nonatomic) IBOutlet UIButton *zipButton;
@property (weak, nonatomic) IBOutlet UIButton *unzipButton;
@property (weak, nonatomic) IBOutlet UIButton *hasPasswordButton;
@property (weak, nonatomic) IBOutlet UIButton *resetButton;
@property (weak, nonatomic) IBOutlet UILabel *file1;
@property (weak, nonatomic) IBOutlet UILabel *file2;
@property (weak, nonatomic) IBOutlet UILabel *file3;
@property (weak, nonatomic) IBOutlet UILabel *info;
@property (copy, nonatomic) NSString *samplePath;
@property (copy, nonatomic) NSString *zipPath;

@end

@implementation ViewController

#pragma mark - Life Cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    
    _samplePath = [[NSBundle mainBundle].bundleURL
                   URLByAppendingPathComponent:@"Sample Data"
                   isDirectory:YES].path;
    NSLog(@"Sample path: %@", _samplePath);
    
    [self resetPressed:_resetButton];
}

#pragma mark - IBAction
- (IBAction)zipPressed:(id)sender {
    _zipPath = [self tempZipPath];
    NSLog(@"Zip path: %@", _zipPath);
    NSString *password = _passwordField.text;
    BOOL success = [SSZipArchive createZipFileAtPath:_zipPath
                             withContentsOfDirectory:_samplePath
                                 keepParentDirectory:NO
                                    compressionLevel:-1
                                            password:password.length > 0 ? password : nil
                                                 AES:YES
                                     progressHandler:nil];
    if (success) {
        NSLog(@"Success zip");
        self.info.text = @"Success zip";
        _unzipButton.enabled = YES;
        _hasPasswordButton.enabled = YES;
    } else {
        NSLog(@"No success zip");
        self.info.text = @"No success zip";
    }
    _resetButton.enabled = YES;
}

- (IBAction)unzipPressed:(id)sender {
    if (!_zipPath) {
        return;
    }
    NSString *unzipPath = [self tempUnzipPath];
    NSLog(@"Unzip path: %@", unzipPath);
    if (!unzipPath) {
        return;
    }
    NSString *password = _passwordField.text;
    BOOL success = [SSZipArchive unzipFileAtPath:_zipPath
                                   toDestination:unzipPath
                              preserveAttributes:YES
                                       overwrite:YES
                                  nestedZipLevel:0
                                        password:password.length > 0 ? password : nil
                                           error:nil
                                        delegate:nil
                                 progressHandler:nil
                               completionHandler:nil];
    if (success) {
        NSLog(@"Success unzip");
        self.info.text = @"Success unzip";
    } else {
        NSLog(@"No success unzip");
        self.info.text = @"No success unzip";
        return;
    }
    NSError *error = nil;
    NSMutableArray<NSString *> *items = [[[NSFileManager defaultManager]
                                          contentsOfDirectoryAtPath:unzipPath
                                          error:&error] mutableCopy];
    if (error) {
        return;
    }
    [items enumerateObjectsUsingBlock:^(NSString * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        switch (idx) {
            case 0: {
                self.file1.text = obj;
                break;
            }
            case 1: {
                self.file2.text = obj;
                break;
            }
            case 2: {
                self.file3.text = obj;
                break;
            }
            default: {
                NSLog(@"Went beyond index of assumed files");
                break;
            }
        }
    }];
    _unzipButton.enabled = NO;
}

- (IBAction)hasPassword:(id)sender {
    if (!_zipPath) {
        return;
    }
    BOOL success = [SSZipArchive isFilePasswordProtectedAtPath:_zipPath];
    if (success) {
        NSLog(@"Yes, it's password protected.");
        self.info.text = @"Yes, it's password protected.";
    } else {
        NSLog(@"No, it's not password protected.");
        self.info.text = @"No, it's not password protected.";
    }
}

- (IBAction)resetPressed:(id)sender {
    _file1.text = @"";
    _file2.text = @"";
    _file3.text = @"";
    _info.text = @"";
    _zipButton.enabled = YES;
    _unzipButton.enabled = NO;
    _hasPasswordButton.enabled = NO;
    _resetButton.enabled = NO;
}

#pragma mark - Private
- (NSString *)tempZipPath {
    NSString *path = [NSString stringWithFormat:@"%@/%@.zip",
                      NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES)[0],
                      [NSUUID UUID].UUIDString];
    return path;
}

- (NSString *)tempUnzipPath {
    NSString *path = [NSString stringWithFormat:@"%@/%@",
                      NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES)[0],
                      [NSUUID UUID].UUIDString];
    NSURL *url = [NSURL fileURLWithPath:path];
    NSError *error = nil;
    [[NSFileManager defaultManager] createDirectoryAtURL:url
                             withIntermediateDirectories:YES
                                              attributes:nil
                                                   error:&error];
    if (error) {
        return nil;
    }
    return url.path;
}

@end

File: CoresRetro/RetroArch/ZipArchive/Example/SwiftExample/ViewController.swift
//
//  ViewController.swift
//  SwiftExample
//
//  Created by Sean Soper on 10/23/15.
//
//

import UIKit

#if UseCarthage
    import ZipArchive
#else
    import SSZipArchive
#endif

class ViewController: UIViewController {

    @IBOutlet weak var passwordField: UITextField!
    @IBOutlet weak var zipButton: UIButton!
    @IBOutlet weak var unzipButton: UIButton!
    @IBOutlet weak var hasPasswordButton: UIButton!
    @IBOutlet weak var resetButton: UIButton!

    @IBOutlet weak var file1: UILabel!
    @IBOutlet weak var file2: UILabel!
    @IBOutlet weak var file3: UILabel!
    @IBOutlet weak var info: UILabel!

    var samplePath: String!
    var zipPath: String?

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        samplePath = Bundle.main.bundleURL.appendingPathComponent("Sample Data").path
        print("Sample path:", samplePath!)
        
        resetPressed(resetButton)
    }

    // MARK: IBAction

    @IBAction func zipPressed(_: UIButton) {
        zipPath = tempZipPath()
        print("Zip path:", zipPath!)
        let password = passwordField.text ?? ""

        let success = SSZipArchive.createZipFile(atPath: zipPath!,
                                                 withContentsOfDirectory: samplePath,
                                                 keepParentDirectory: false,
                                                 compressionLevel: -1,
                                                 password: !password.isEmpty ? password : nil,
                                                 aes: true,
                                                 progressHandler: nil)
        if success {
            print("Success zip")
            info.text = "Success zip"
            unzipButton.isEnabled = true
            hasPasswordButton.isEnabled = true
        } else {
            print("No success zip")
            info.text = "No success zip"
        }
        resetButton.isEnabled = true
    }

    @IBAction func unzipPressed(_: UIButton) {
        guard let zipPath = self.zipPath else {
            return
        }

        guard let unzipPath = tempUnzipPath() else {
            return
        }
        print("Unzip path:", unzipPath)
        
        let password = passwordField.text ?? ""
        let success: Bool = SSZipArchive.unzipFile(atPath: zipPath,
                                                   toDestination: unzipPath,
                                                   preserveAttributes: true,
                                                   overwrite: true,
                                                   nestedZipLevel: 1,
                                                   password: !password.isEmpty ? password : nil,
                                                   error: nil,
                                                   delegate: nil,
                                                   progressHandler: nil,
                                                   completionHandler: nil)
        if success != false {
            print("Success unzip")
            info.text = "Success unzip"
        } else {
            print("No success unzip")
            info.text = "No success unzip"
            return
        }

        var items: [String]
        do {
            items = try FileManager.default.contentsOfDirectory(atPath: unzipPath)
        } catch {
            return
        }

        for (index, item) in items.enumerated() {
            switch index {
            case 0:
                file1.text = item
            case 1:
                file2.text = item
            case 2:
                file3.text = item
            default:
                print("Went beyond index of assumed files")
            }
        }

        unzipButton.isEnabled = false
    }

    @IBAction func hasPasswordPressed(_: UIButton) {
        guard let zipPath = zipPath else {
            return
        }
        let success = SSZipArchive.isFilePasswordProtected(atPath: zipPath)
        if success {
            print("Yes, it's password protected.")
            info.text = "Yes, it's password protected."
        } else {
            print("No, it's not password protected.")
            info.text = "No, it's not password protected."
        }
    }

    @IBAction func resetPressed(_: UIButton) {
        file1.text = ""
        file2.text = ""
        file3.text = ""
        info.text = ""
        zipButton.isEnabled = true
        unzipButton.isEnabled = false
        hasPasswordButton.isEnabled = false
        resetButton.isEnabled = false
    }

    // MARK: Private

    func tempZipPath() -> String {
        var path = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true)[0]
        path += "/\(UUID().uuidString).zip"
        return path
    }

    func tempUnzipPath() -> String? {
        var path = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true)[0]
        path += "/\(UUID().uuidString)"
        let url = URL(fileURLWithPath: path)

        do {
            try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true, attributes: nil)
        } catch {
            return nil
        }
        return url.path
    }
    
}
